### DISCUSSION TITLE
How is High Availablility (HA) supposed to work?

### DISCUSSION THREAD

**Discussion by jerrac** [keycloak/keycloak#27301] [Q&A]
I've spent a week or two, off and on, attempting to figure out how to make HA work with Keycloak. Long story short, I haven't succeeded, and my reading of the docs and various search results hasn't gotten me up and running.

The only docs I have found for HA and Keycloak 22+ are [here](https://www.keycloak.org/guides#high-availability) which seems to describe active/passive HA between physical locations (dependent on Kubernetes for some reason...), and [here](https://www.keycloak.org/server/caching) which I think implies that active/active HA is create by making sure Infinispan is working. I saw some posts that seem to indicate active/active was supported in older versions, but then things changed when Quarkus entered the picture. I also found [this discussion](https://github.com/infinispan/infinispan/discussions/9885) that says embedded Infinispan does not support rolling upgrades. 

I tried to figure out how to set up Infinispan server and point Keycloak at it. Searching led me to [this in preview doc](https://www.keycloak.org/high-availability/connect-keycloak-to-external-infinispan) that I tried to adapt to my Docker Swarm setup. 

My goal is single site active/active HA with 3 Keycloak instances load balanced so that if one goes down, the load balancer will send traffic to the others, and users will not notice anything. This would include rolling upgrades.

So I'd like to ask a few basic questions to make sure I'm not messing up the basics before I go and post about specific configurations and errors.

A) Does Keycloak support active/active HA?

B) For each Keycloak instance/replica, do I point them at the same database, or do they each have their own db?

C) Does Keycloak itself support rolling upgrades?

D) Infinispan is the means by which Keycloak operates as a cluster, right?

E) Has anyone else succeeded in setting up active/active HA on Docker Swarm?

Thanks in advance!

**Comment by eestolano** [comment]
A) yes
B) they use the same database
C) no rolling upgrades see https://github.com/keycloak/keycloak/discussions/21067
D) yes, Infinispan
E) I've tried it with bare Docker, but not Docker Swarm

We currently run Keycloak with active/active HA using Infinispan. Each instance is a Docker container on a separate VM, without Docker Swarm.

Getting the Infinispan transport stack setup right was the hardest part. The available documentation isn't great. There are [some transport stack options mentioned here](https://www.keycloak.org/server/caching#_transport_stacks) but none of them were suitable for our situation.

Eventually I found [this gist](https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80) with an example configuration that uses the JDBC_PING protocol, which does Infinispan cluster discovery via the Keycloak database itself. We did have to make a few tweaks to get it to work for us, mainly:
* That example is written for postgres, and our Keycloak DB is mysql, so I had to modify it accordingly.
* I added `<TCP external_addr="${env.INTRANET_IP}" />` to the custom stack (between lines 9 and 10 in the gist), and had each container set `INTRANET_IP` to the address at which other Keycloak instances can reach it.
* We had to make sure to map/expose the default infinispan port (port 7800).

**Reply by jerrac** [reply]
Thanks for the reply. :)

I ended up trying TCPPING and setting Infinispan up via standalone docker compose projects on separate vms. I haven't been able to get it to the point I trust it to be stable. Way too many logs about one instance or another leaving and rejoining the cluster. Even when they're using `network_mode: host` and I've triple checked the various ports on the firewalls...

I also never figured out how to properly get Keycloak to point the distributed caches at the cluster securely. 'Course, it was still in my Swarm cluster, so now I'm wondering if there was something odd going on with the connections due to the overlay network... Hmm... The overlay network doesn't support multicast, but that shouldn't be needed when Infinispan is outside the cluster, right? 

Um, so, looking a little closer at that gist, is it clustering by having Keycloak join the Infinispan cluster, instead of telling Keycloak to point the distributed cache's at the Infinispan cluster via host/username/password?

Is that what you did?

**Reply by eestolano** [reply]
Yes, the Keycloak instances are joining the Infinispan cluster. They actually _are_ the entire Infinispan cluster; we're not running any additional/separate Infinispan services.

(This does have the downside that any full, non-rolling shutdown--such as for an upgrade--blows away the distributed cache. That's not a problem for us, but ymmv.)

**Reply by jerrac** [reply]
Oohh. The embedded cluster. I actually did have that working, at least sortof, for a while. Then I read infinispan doesn't support rolling upgrades for the embedded version and started trying to get the remote server version working.

Makes me wonder if there's a way to have the embedded instances join the server instances... I might try that later.
